### ISSUE TITLE
Use LDR_CNTRL in AIX to use 64KB page sizes.

### ISSUE THREAD

**Issue by rmnattas** [eclipse-openj9/openj9#19052]
We can improve performance on AIX by setting LDR_CNTRL to use 64KB pages by default.

`LDR_CNTRL=TEXTPSIZE=64K@DATAPSIZE=64K@STACKPSIZE=64K@SHMPSIZE=64K`

This is proposed to OpenJDK https://bugs.openjdk.org/browse/JDK-8252802 but rejected.

- [ ] JDK8: https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/739
- [ ] JDK11: https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/760
- [ ] JDK17: https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/330
- [ ] JDK21: https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/134
- [ ] JDK22: https://github.com/ibmruntimes/openj9-openjdk-jdk22/pull/30
- [ ] JDKNext: https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/749

**Comment by rmnattas** [comment]
Tested couple workloads locally and it may have helped somewhat or stayed the same, with not significant impact on memory usage (positive or negative), hence no significant implication. Also pLinux already uses 64KB everywhere by default.

**Comment by pshipton** [comment]
That doesn't seem like a strong justification to go ahead with the change and the potential risks. Everything forked from the JVM will have the same settings.

**Comment by pshipton** [comment]
Pls run internal builds with all the test suites.

**Comment by pshipton** [comment]
Besides the testing, the change should include an option to control the behavior, so a user or dev can turn it off for diagnostics, if it's causing problems, or the change isn't desired. An example similar option is `-XX:[+/-]UseZlibNX`.
https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/src/java.base/unix/native/libjli/java_md.c#L390

**Comment by zl-wang** [comment]
that sounds reasonable. we only set it when -XX:-YYY is not set.

**Comment by keithc-ca** [comment]
Is anything happening to progress this? Or should this issue and the referenced pull requests be closed?

**Comment by zl-wang** [comment]
@keithc-ca  can you help this out? otherwise, we will come back to this later

**Comment by keithc-ca** [comment]
OpenJDK pointed out that `MALLOCOPTIONS` must be set *before* a process is created; is the same true for `LDR_CNTRL`?

**Comment by zl-wang** [comment]
yes ... as always, need to set in the launcher (as using NX zlib).
it could be a little involved: we need to evaluate
1) memory footprint implication 
2) performance benefit evaluation
3) possible remedy/mitigation to increased memory footprint

**Comment by pshipton** [comment]
Afaik this bogged down on my request to introduce an option to control the behavior. 
https://github.com/eclipse-openj9/openj9/issues/19052#issuecomment-1983522399

**Comment by keithc-ca** [comment]
That wasn't my question. If it must be set before the process is created, anything we might do in the launcher is too late.

**Comment by keithc-ca** [comment]
I can help (the reference to `UseZlibNX` is useful), but there's no point in code changes that can't benefit the process after it has started.

**Comment by pshipton** [comment]
I think the launcher forks a new process.

**Comment by zl-wang** [comment]
UseZlibNX sets up LIBPATH ... it can take effect for the java run.  why not for LDR_CNTRL?

**Comment by keithc-ca** [comment]
I starts a new thread to run the JVM, but I don't see `fork()` anywhere in the launcher, so there's no new process.

**Comment by pshipton** [comment]
See https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/openj9/src/java.base/unix/native/libjli/java_md_solinux.c#L520

**Comment by pshipton** [comment]
Perhaps "fork" was the wrong term to use.

**Comment by keithc-ca** [comment]
That's `execve`, not `fork` (same process, new executable).

**Comment by pshipton** [comment]
Afaik it's working for the `MALLOCOPTIONS` we set in there.

**Comment by keithc-ca** [comment]
Maybe the documentation for `MALLOCOPTIONS` is imprecise; the requirement is that the environment be set at the time `execve()` (or similar) is called?

I'll begin work on this.

**Comment by zl-wang** [comment]
I believed whoever documented `MALLOCOPTIONS` for AIX was unaware of `execve` (and its brothers)  new-process image (it  should be taken as a start-up too from env setting perspective, since env is explicitly passed in).  exactly because `malloc()` on AIX was dog slow in multi-thread program, we used `MALLOCOPTIONS`  in launcher (since JDK8) to make it better. obviously, it took effect. I remembered we needed to set it to a specific value, otherwise it can cause excessive memory consumption.

**Comment by keithc-ca** [comment]
Any opinions on what the controlling command-line options should be? My prototype uses `-XX:[+-]setLDR_CNTRL`, but that's easily changed.

**Comment by zl-wang** [comment]
maybe  `-XX:[+-]mediumPageByDefault`  or something like that to try  not that AIX-specific

**Comment by keithc-ca** [comment]
This whole thing is AIX-specific. I don't think `ByDefault` is appropriate because the options are explicitly specifying whether or not to adjust page sizes - overriding any default. I'm open to something more general like `-XX:[+-]MediumPageSize`. @pshipton Do you have any thoughts or suggestions?

**Comment by zl-wang** [comment]
then, what `-XX:-Option ` would do in this case, since it is not in effect by default anyway? you need the +Option to opt in explicitly. The name `MediumPageSize ` is fine with me ... maybe `UseMediumPages` is better?  the former sounds like a query for the size

**Comment by keithc-ca** [comment]
Our practice is to have both enabling and disabling options, independent of which is the default. That way you can always append an option to specify what you want no matter what the default or whether preceding options specified something else.

I'll go with `UseMediumPageSize` unless there are objections.

**Comment by pshipton** [comment]
As long as there is an option to turn it off in case it breaks something, I'm good with it.

**Comment by keithc-ca** [comment]
Change for all versions and document update are merged: Closing.
